### ISSUE TITLE
Windows CUDA builds are failing in HUD

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#54382]
pytorch_windows_vs2019_py36_cuda10.1_build, pytorch_windows_vs2019_py36_cuda11.1_build, and pytorch_windows_vs2019_py36_cuda11.2_build are failing in the HUD.

Snippet of the failure:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include\cmath(124): error: calling a __host__ function("__copysignf") from a __host__ __device__ function("c10::guts::detail::apply_impl< ::at::native::AUnaryFunctor< ::>  &,     ::std::tuple<float >  &, (unsigned long long)0ull > ") is not allowed

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include\cmath(141): error: calling a __host__ function("__floorf") from a __host__ __device__ function("c10::guts::detail::apply_impl< ::at::native::BUnaryFunctor< ::>  &,     ::std::tuple<float >  &, (unsigned long long)0ull > ") is not allowed

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include\cmath(124): error: calling a __host__ function("__copysignf") from a __host__ __device__ function("c10::guts::detail::apply_impl< ::at::native::BUnaryFunctor< ::>  &,     ::std::tuple<float >  &, (unsigned long long)0ull > ") is not allowed

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include\cmath(141): error: calling a __host__ function("__floorf") from a __host__ __device__ function("c10::guts::detail::apply_impl< :: &,     ::std::tuple<float, float >  &, (unsigned long long)0ull, (unsigned long long)1ull > ") is not allowed

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include\cmath(124): error: calling a __host__ function("__copysignf") from a __host__ __device__ function("c10::guts::detail::apply_impl< :: &,     ::std::tuple<float, float >  &, (unsigned long long)0ull, (unsigned long long)1ull > ") is not allowed

48 errors detected in the compilation of "C:/Users/circleci/project/build/win_tmp/bin/.tmpsHP08a/tmpxft_00000d98_00000000-7_BinaryMulDivKernel.cpp1.ii".
BinaryMulDivKernel.cu
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_BinaryMulDivKernel.cu.obj
C:/Jenkins/Miniconda3/Library/bin/cmake.exe -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_BinaryMulDivKernel.cu.obj
CMake Error at torch_cuda_generated_BinaryMulDivKernel.cu.obj.Release.cmake:281 (message):
  Error generating file
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_BinaryMulDivKernel.cu.obj
```

Surprisingly, the issue begins at PR that seems unrelated (a small docs change). 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @malfet @seemethere @walterddr @ngimel @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

**Comment by peterjc123** [comment]
According to early reports, should be caused by the vs toolchain update of the circleci image. The MSVC update includes an update of the file `cmath`, which forwards all the calls of `floorf` and the `ceilf` to the implementation using instrinics, which cannot be used with NVCC.

**Comment by skyline75489** [comment]
I was having this discussion with my team internally at MS about this. I wonder what is the strategy of updating VS toolchain at the moment? Can we just freeze the toolchain version since i do see obvious gain from updating toolchain frequently.

**Comment by peterjc123** [comment]
Occurrences of the related functions in ATen
```cpp
std::copysign:
aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:          floordiv = std::copysign(scalar_t(0), a * inv_b);
aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:          floordiv = std::copysign(scalar_t(0), a / b);
aten/src/ATen/native/cuda/Math.cuh:    return std::copysign(static_cast<scalar_t>(INFINITY), -x);

std::ceil:
aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:#define END_IND(a,b,c) (int)std::ceil((float)((a + 1) * c) / b)
aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:  return (int)std::ceil((float)((a + 1) * c) / b);
aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:  return (int)std::ceil((float)((a + 1) * c) / b);
aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu:  return (int)std::ceil((float)((a + 1) * c) / b);
aten/src/ATen/native/cuda/RangeFactories.cu:      size_d = std::ceil(static_cast<double>(end.to<accscalar_t>() - start.to<accscalar_t>())
aten/src/ATen/native/cuda/RangeFactories.cu:      size_d = std::ceil(static_cast<double>(end.to<double>() - start.to<double>())
aten/src/ATen/native/cuda/ReflectionPad.cu:    (int) std::ceil(output_plane_size/256.0), nplane, nbatch);
aten/src/ATen/native/cuda/ReflectionPad.cu:    (int) std::ceil(output_plane_size/256.0), nplane, nbatch);
aten/src/ATen/native/cuda/UnaryFractionKernels.cu:// We manually overload ceil because std::ceil does not work with std::complex types.
aten/src/ATen/native/cuda/UnaryFractionKernels.cu:  return std::ceil(a);
aten/src/ATen/native/cuda/UnaryFractionKernels.cu:  return std::complex<T>(std::ceil(v.real()), std::ceil(v.imag()));

std::floor:
aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:#define START_IND(a,b,c) (int)std::floor((float)(a * c) / b)
aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:  return (int)std::floor((float)(a * c) / b);
aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:  return (int)std::floor((float)(a * c) / b);
aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu:  return (int)std::floor((float)(a * c) / b);
aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:          floordiv = std::floor(div);
aten/src/ATen/native/cuda/BinaryMulDivKernel.cu:          floordiv = std::floor(div);
aten/src/ATen/native/cuda/UnaryFractionKernels.cu:// We manually overload floor because std::floor does not work with std::complex types.
aten/src/ATen/native/cuda/UnaryFractionKernels.cu:  return std::floor(a);
aten/src/ATen/native/cuda/UnaryFractionKernels.cu:  return std::complex<T>(std::floor(v.real()), std::floor(v.imag()));
```

**Comment by peterjc123** [comment]
Local test results:
```
std::ceil(float) fail -> std::ceilf(float) pass
std::ceil(double) pass

std::floor(float) fail -> std::floorf(float) pass
std::floor(double) pass

std::copysign(float, float) fail -> std::copysignf(float, float) pass
std::copysign(double, double) pass
```

**Comment by peterjc123** [comment]
https://github.com/pytorch/pytorch/pull/54386 is a PoC for the fix. You'll just need to apply it to all the related sources to make the build pass.

**Comment by peterjc123** [comment]
Other simple ways to workground this problem:
1. Downgrade VS toolchain
2. Attach a copy of the old `<cmath>` file and add the include path for it in NVCC

**Comment by malfet** [comment]
Downgrading the toolchain sounds like a reasonable solution to me.

**Comment by malfet** [comment]
Last working version were  MSVC 19.28.29337.0 , not sure how to roll back to that one, trying to roll back to 14.27 in #54399

**Comment by mszhanyi** [comment]
I think we should always pin VC to a stable version and upgrade it  when it's necessary.
Thus, it won't break CI.

**Comment by mszhanyi** [comment]
There's 2 VCs in the image 
![image](https://user-images.githubusercontent.com/16190118/111965390-a651c580-8b30-11eb-9cdb-d2e824a625d4.png)

**Comment by peterjc123** [comment]
> There's 2 VCs in the image of windows-server-2019-vs2019:stable
> ![image](https://user-images.githubusercontent.com/16190118/111965390-a651c580-8b30-11eb-9cdb-d2e824a625d4.png)

14.16 is VS2017. I don't think we could use that.

> I think we should always pin VC to a stable version and upgrade it when it's necessary.
> Thus, it won't break CI.

Yeah, this is a good idea. Actually, we could try to use the docker container, at least for the build phase. And previous work is done in https://github.com/pytorch/pytorch/pull/45376.

**Comment by peterjc123** [comment]
@malfet @mszhanyi 

> 2\. Attach a copy of the old `<cmath>` file and add the include path for it in NVCC (#54391)

What do you think about this option? I think it's pretty easy to apply and may be the eaiest option we may take.

**Comment by malfet** [comment]
> > 2. Attach a copy of the old `<cmath>` file and add the include path for it in NVCC (#54391)
> 
> What do you think about this option? I think it's pretty easy to apply and may be the eaiest and most convenient option we may take. What's more, it really solves the problem for end users. Imagine how complicated it is for a user to install an old toolchain and activate it on his own machine.

Imo this approach is problematic for the following reasons:
 - It lulls developers into believing that something works when it does not
 - It makes it impossible to compile code on standalone machine
 - It soft-embedds header under a different license agreement into the codebase
Imo best solution is similar to what you've started in #54386: create CUDA-friendly `c10::ceil`, `c10::floor` and `c10:copysign` aliases.

**Comment by malfet** [comment]
I wonder if https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDAMathCompat.h is already accomplishes that...

**Comment by peterjc123** [comment]
> I wonder if https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDAMathCompat.h is already accomplishes that...

Nope, in this header, it only takes the functions from CRT, but the CRT ones are also broken.

**Comment by peterjc123** [comment]
>     * It makes it impossible to compile code on standalone machine

You mean offline compiling right? Well in this case, downgrading won't work, too.


>     * It soft-embedds header under a different license agreement into the codebase

STL uses the Apache license, so it should be okay.



>     * Imo best solution is similar to what you've started in #54386: create CUDA-friendly `c10::ceil`, `c10::floor` and `c10:copysign` aliases.

It sounds reasonable, but requires too much work. I don't think this is our first choice, given that we have so many other choices at hand that is much easier.

**Comment by malfet** [comment]
> > ```
> > * It makes it impossible to compile code on standalone machine
> > ```
> 
> You mean offline compiling right? Well in this case, downgrading won't work, too.

In my opinion, there is a difference between installing an older version of toolkit and downloading an external header from the internet every time CMake is invoked.

> STL uses the Apache license, so it should be okay.

In that case, perhaps we should make it a PyTorch submodule.


> > ```
> > * Imo best solution is similar to what you've started in #54386: create CUDA-friendly `c10::ceil`, `c10::floor` and `c10:copysign` aliases.
> > ```
> 
> It sounds reasonable, but requires too much work. I don't think this is our first choice, given that we have so many other choices at hand that is much easier.

Completely agree with you here. Do you mind if I'll iterate over your workaround to make it a part of CI (that just overrides cmath.h) and then will work on a longer path to make PyTorch codebase compatible with latest VS2019

**Comment by peterjc123** [comment]
> Completely agree with you here. Do you mind if I'll iterate over your workaround to make it a part of CI (that just overrides cmath.h) and then will work on a longer path to make PyTorch codebase compatible with latest VS2019

Sure, no problem.

**Comment by ptrblck** [comment]
This might be related to: https://github.com/microsoft/onnxruntime/issues/6927

A workaround could be to use: `--msvc_toolset=14.27`, as this functional regression seems to be introduced in toolset `14.28`.

**Comment by mszhanyi** [comment]
It's caused by  regression of VS. We've pinned the VS version for windows build

**Comment by sylveon** [comment]
Oops, I'm the one that broke that in https://github.com/microsoft/STL/pull/1336

**Comment by sylveon** [comment]
Fixed in 16.10 preview 4

**Comment by StephanTLavavej** [comment]
FYI, VS 2019 16.9.7 was [released](https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes-v16.9#--visual-studio-2019-version-1697-) last week, and contains the microsoft/STL fix for this bug (which is specifically mentioned in the release notes: "Fixed an issue preventing developers to build PyTorch.")

Apologies for the headache. We're trying to be more careful about intrinsic usage in the future, but if you notice any problems with STL / CUDA scenarios, please let us know so we can fix them quickly.
